pkg/runtime/log/resource.go

```diff
@@ -14,12 +14,102 @@
 package log
 
 import (
+	"strings"
+
 	"github.com/go-logr/logr"
 
 	"github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
 	acktypes "github.com/aws-controllers-k8s/runtime/pkg/types"
 )
 
+// ResourceLogger is a wrapper around a logr.Logger that writes log messages
+// about resources involved in a controller loop. It implements
+// `pkg/types.AWSLogger`
+type ResourceLogger struct {
+	log        logr.Logger
+	res        acktypes.AWSResource
+	blockDepth int
+}
+
+// WithValues adapts the internal logger with a set of additional values
+func (rl *ResourceLogger) WithValues(
+	values ...interface{},
+) {
+	rl.log = rl.log.WithValues(values...)
+}
+
+// Debug writes a supplied log message about a resource that includes a set of
+// standard log values for the resource's kind, namespace, name, etc
+func (rl *ResourceLogger) Debug(
+	msg string,
+	additionalValues ...interface{},
+) {
+	AdaptResource(rl.log, rl.res, additionalValues...).V(1).Info(msg)
+}
+
+// Info writes a supplied log message about a resource that includes a
+// set of standard log values for the resource's kind, namespace, name, etc
+func (rl *ResourceLogger) Info(
+	msg string,
+	additionalValues ...interface{},
+) {
+	AdaptResource(rl.log, rl.res, additionalValues...).V(0).Info(msg)
+}
+
+// Enter logs an entry to a function or code block
+func (rl *ResourceLogger) Enter(
+	name string, // name of the function or code block we're entering
+	additionalValues ...interface{},
+) {
+	rl.blockDepth++
+	depth := strings.Repeat(">", rl.blockDepth)
+	msg := depth + " " + name
```

Comment:
We can avoid doing unnecessary allocations/ops here

```go
        if rl.log.V(1).Enabled() {
	    rl.blockDepth++
	    depth := strings.Repeat(">", rl.blockDepth)
	    msg := depth + " " + name
            AdaptResource(rl.log, rl.res, additionalValues...).V(1).Info(msg)
        }
```

Reply:
sweet. nice suggestion! :+1: